lib/style/defs.ex

```diff
@@ -41,60 +41,107 @@ defmodule Styler.Style.Defs do
 
   @behaviour Styler.Style
 
+  alias Styler.Style
   alias Styler.Zipper
 
   # a def with no body like
   #
   #  def example(foo, bar \\ nil)
   #
   def run({{def, meta, [head]}, _} = zipper, ctx) when def in [:def, :defp] do
+    {_fn_name, head_meta, _children} = head
+    first_line = meta[:line]
+    last_line = head_meta[:closing][:line]
+
+    comments =
+      if first_line == last_line do
+        ctx.comments
+      else
+        Style.displace_comments(ctx.comments, first_line..last_line)
+      end
+
     # There won't be any defs deeper in here, so lets skip ahead if we can
-    {:skip, Zipper.replace(zipper, {def, meta, [flatten_head(head, meta[:line])]}), ctx}
+    head = flatten_head(head, meta[:line])
+    {:skip, Zipper.replace(zipper, {def, meta, [head]}), %{ctx | comments: comments}}
   end
 
   # all the other kinds of defs!
   def run({{def, def_meta, [head, body]}, _} = zipper, ctx) when def in [:def, :defp] do
-    def_start_line = def_meta[:line]
-    # order matters here! the end of the def is where the `do`s line is - but only if there's a do end block.
-    # otherwise it's just where the end of the (def) expression is.
-    def_end_line = (def_meta[:do] || def_meta[:end_of_expression])[:line]
+    {style, def_line, do_line, end_line} =
+      if def_meta[:do] do
+        # This is a def with a do block, like
+        #
+        #  def example(foo, bar \\ nil) do
+        #    :ok
+        #  end
+        #
+        def_line = def_meta[:line]
+        do_line = def_meta[:do][:line]
+        end_line = def_meta[:end][:line]
+        {:block, def_line, do_line, end_line}
+      else
+        # This is a def with a keyword do, like
+        #
+        #  def example(foo, bar \\ nil), do: :ok
+        #
+        [{{:__block__, do_meta, [:do]}, {_, body_meta, _}}] = body
+        def_line = def_meta[:line]
+        do_line = do_meta[:line]
+        end_line = body_meta[:closing][:line] || do_meta[:line]
+        {:keyword, def_line, do_line, end_line}
+      end
 
-    if def_start_line == def_end_line do
+    if def_line == end_line do
```

Comment:
nit:
hmm, could cut down on `if block else` in here by getting just the end line above, and leaving the `do_line` determination inside of the `if block`
like:
```elixir
def_line = def_meta[:line]
end_line =
  if line = def_meta[:do][:line] || def_meta[:end_of_expression][:line] do
    line
  else
    [{{:__block__, do_meta, [:do]}, {_, body_meta, _}}] = body
    body_meta[:closing][:line] || do_meta[:line]
  end

cond do
  def_line == end_line -> :skip
  do_line = def_meta[:do][:line] ->
    # we're in a def do end
    ... 
    {:skip, ...}
  true ->
    # we're in a keyword
    ...
    {:skip, ...}
end
```
sorry, had to write that out to see how it compares. shrug. i think it is fine to not do the `{node, comments} = if ..` just for the sake of only having one skip statement.

Reply:
>i think it is fine to not do the {node, comments} = if .. just for the sake of only having one skip statement.

this sentence hard to read. let me try again.

i think only having one return site, and so matching {node, comments} out of an if statement, is maybe harder to follow than just having each branch of the statement have its own return value

Reply:
Yeah I tried it both ways and landed where I did because I thought there was more duplication between the branches than there actually turned out to be, because both branches are basically calling the same functions but with slightly different parameters.

I'll give it a try like you're sketching out above and see how it looks.

Reply:
Ok yeah I think I do like it better with a `cond` so let's go with that. ❤️